src/G2/Equiv/Verifier.hs

```diff
@@ -329,28 +470,47 @@ obligationWrap obligations =
     then Nothing
     else Just $ ExtCond (App (Prim Not TyUnknown) conj) True
 
+totalName :: [DT.Text] -> Name -> Bool
+totalName texts (Name t _ _ _) = t `elem` texts
+
 checkRule :: Config ->
              State t ->
              Bindings ->
+             [DT.Text] ->
              RewriteRule ->
              IO (S.Result () ())
-checkRule config init_state bindings rule = do
+checkRule config init_state bindings total rule = do
   let (rewrite_state_l, bindings') = initWithLHS init_state bindings $ rule
       (rewrite_state_r, bindings'') = initWithRHS init_state bindings' $ rule
       -- convert from State t to StateET
       e_l = exprExtract rewrite_state_l
       e_r = exprExtract rewrite_state_r
+      -- Tick wrapping for recursive functions
+      h_l = expr_env rewrite_state_l
+      h_r = expr_env rewrite_state_r
+      -- handle corecursion, not just direct recursion
+      wrap_l = wrapAllRecursion (G.getCallGraph h_l) h_l
+      wrap_r = wrapAllRecursion (G.getCallGraph h_r) h_r
+      total_names = filter (totalName total) (map idName $ ru_bndrs rule)
+      -- TODO do I still need the HashSet usage elsewhere?
+      total_hs = foldr HS.insert HS.empty total_names
+      EquivTracker et m _ = emptyEquivTracker
+      start_equiv_tracker = EquivTracker et m total_hs
       rewrite_state_l' = rewrite_state_l {
-                           track = emptyEquivTracker
+                           track = start_equiv_tracker
                          , curr_expr = CurrExpr Evaluate $ tickWrap $ e_l
+                         , expr_env = E.mapWithKey wrap_l h_l
                          }
       rewrite_state_r' = rewrite_state_r {
-                           track = emptyEquivTracker
+                           track = start_equiv_tracker
                          , curr_expr = CurrExpr Evaluate $ tickWrap $ e_r
+                         , expr_env = E.mapWithKey wrap_r h_r
```

Comment:
Looks like a lot of lines throughout checkRule that are copy+pasted to apply to both left and right hand side.  Maybe make a new function that does all the initialization for a state, and then we just call that function on both the LHS and RHS?